design/824-add-nchains-option-to-mcmc/discussion.qmd

```diff
@@ -0,0 +1,1202 @@
+---
+title: "Design/discussion document for multiple MCMC chains"
+format: html
+embed-resources: TRUE
+editor: visual
+---
+
+```{r}
+#| eval: TRUE
+#| label: set-up
+#| echo: FALSE
+
+suppressPackageStartupMessages({
+  library(crmPack)
+  library(knitr)
+  library(kableExtra)
+  library(mvtnorm)
+  library(checkmate)
+  library(dplyr)
+  library(rjags)
+})
+
+# The Validate class needs to be defined in the Quarto document.  Perhaps because it's not exported from `crmPack`?
+Validate <- setRefClass(
+  Class = "Validate",
+  fields = list(msg = "character"),
+  methods = list(
+    check = function(test, string = "") {
+      "Check whether the \\code{test} is \\code{TRUE}; if so, return \\code{NULL}.
+      Otherwise, add the \\code{string} message into the cumulative messages vector \\code{msg}."
+      assert_flag(test)
+      assert_string(string)
+      if (test) {
+        NULL
+      } else {
+        msg <<- c(msg, string)
+      }
+    },
+    result = function() {
+      "Return either cumulative messages vector \\code{msg}
+      (which contains the error messages from all the checks),
+      or \\code{NULL}, if \\code{msg} is empty (i.e. all the checks were successful)."
+      if (length(msg) > 0) {
+        msg
+      } else {
+        TRUE
+      }
+    }
+  )
+)
+```
+
+# Questions
+
+-   Why does `McmcOptions` treat `burnin` and `n.iter` differently to JAGS? `mcmc` sets the JAGS
+    `n.iter` option to `McmcOptions` `iterations - burnin`.
+-   Should we consider modifying `h_jags_write_model` to use return a string rather than a temporary
+    file? `jags_model` could then use `textConnection`, which removes any interaction with the file
+    system.
+
+# Background
+
+In principle, adding support for multiple MCMC chains to `crmPack` should be straightforward.
+`rjags` supports multiple chains via the `n.chains` argument to `jags.model` and related functions.
+All that it appears we need to do is to add a similar parameter to `mcmc` and we have what we need.
+
+And indeed, this will/should work. But it will not be effective. The reason is that, to test
+convergence effectively, different chains estimating the same model should start from different
+initial positions. If these positions are extreme, so much the better, as it supports the assumption
+that the chains will converge (to the same stable state) regardless of their starting positions.
+
+This is where our difficulties begin. For each model class, `crmPack` provides *constant* starting
+values for the MCMC chain(s). For example, this is the `init` slot for `LogisticLogNormal`:
+
+```{r}
+#| eval: FALSE
+#| label: example-init-slot
+
+init <- function() {
+  list(theta = c(0, 1))
+}
+```
+
+Clearly, what values can be considered extreme depend on the prior distribution of the model. These
```

Comment:
```suggestion
Clearly, what values can be considered extreme depends on the prior distribution of the model. These
```

Reply:
If that's my only spelling mistake, I'm amazed!